### PR TITLE
Fixes Sample Save As exposing editable UI for shared datasets

### PR DESCRIPTION
**Pull request by acarrera42**

### CODE REVIEW

File: app/models/data_import.rb [hunk 1]
```diff
@@ -847,8 +847,13 @@ def store_results(importer=nil, runner=nil, datasource_provider=nil, manual_fiel
       end
 
       # Table.after_create() setted fields that won't be saved to "final" data import unless specified here
-      self.table_name = importer.table.name if importer.success? && importer.table
-      self.table_id   = importer.table.id if importer.success? && importer.table
+      # Hack for Samples 2.0 Save As - Tech debt to clean up with more elegant solution
+      #  If we are import a .carto file do not attach a name to the synchronziation record
```
Comment (tylerparsons) [inline app/models/data_import.rb]: The sync is not given a name, that is true, but the more important consequence of this is that the sync is not linked to the table visualization, correct?  Please update the comment to clarify.

File: services/importer/lib/importer/runner.rb [hunk 2]
```diff
@@ -242,6 +242,71 @@ def file_based_loader_run(job, loader)
...
+            raise CartoDB::Importer2::GenericImportError.new("Failed to create remote visualization")
+          end
+        end
+      end
+
+      def connect_shared_dataset(table_name)
+        if Carto::Synchronization.where(:name => table_name, :user_id => @user.id, :service_name => 'connector').size <= 0
+          # Import the dataset as if the user connected directly to it....
+          member_attributes = {
+            name:                   table_name,
```
Comment (tylerparsons) [inline services/importer/lib/importer/runner.rb:281]: One attribute which is present on all common data syncs yet seems to be missing here is a `url` of the form
```ruby
"http://#{host_name}/user/{#common_data_user.username}/api/v1/sql?q=select+*+from+%22#{table_name}%22&format=shp&filename=#{table_name}"
```
Is this added by the `DataImport`? Or is it never added to the synchronization?

This is not necessarily something that needs to be there, as these syncs will not be run, but we should be aware of the impact of any such discrepancies.

File: spec/connectors/importer_spec.rb [hunk 3]
```diff
@@ -672,22 +672,29 @@
...
       @data_import.success.should eq true
-
+      expect(synchronization.state).to eq 'success'
+      expect(synchronization.name).to be_nil
+
+      # Verify a synchronization record exists per dataset
+      sync = Carto::Synchronization.where(user_id: @user.id, name: 'us_states', service_name: 'connector')
+      expect(sync.any?).to be_true
+      expect(sync.first.state).to eq('success')
+      expect(sync.first.service_item_id).to eq('driver=postgres;channel=postgres_fdw;table=us_states')
```
Comment (tylerparsons) [inline spec/connectors/importer_spec.rb:692]: Add a check for `sync.first.url` against whatever we decide the correct value is per the comment on `connect_shared_dataset`.  `lib/carto/url.rb` may be of use here.

File: spec/connectors/importer_spec.rb [hunk 4]
```diff
@@ -672,22 +672,29 @@
...
 
         @data_import.run_import!
       ensure
         CommonDataSingleton.instance.unstub(:datasets)
       end
 
       @data_import.success.should eq true
-
+      expect(synchronization.state).to eq 'success'
+      expect(synchronization.name).to be_nil
```
Comment (tylerparsons) [inline spec/connectors/importer_spec.rb:685]: Also check that `synchronization.visualization_id` is `Nil`.

File: spec/connectors/importer_spec.rb [hunk 5]
```diff
@@ -739,27 +746,42 @@
...
       expect(user_external_source.visualization_id).to eq user_remote_vis.id
 
       # TODO - Will need to verify load_common_data is not called
-      #expect(CommonDataSingleton.instance).to_not receive(:datasets)
+      # expect(CommonDataSingleton.instance).to_not receive(:datasets)
 
       @data_import.run_import!
       @data_import.success.should eq true
+      expect(synchronization.state).to eq 'success'
+      expect(synchronization.name).to be_nil
```
Comment (tylerparsons) [inline spec/connectors/importer_spec.rb]: Also check that `synchronization.visualization_id` is `Nil`.

File: spec/connectors/importer_spec.rb [hunk 6]
```diff
@@ -705,8 +712,8 @@
       expect(table_vis.export_geom).to be_false
       expect(table_vis.description).to eq('This is my test description')
 
-      # Check the user table was created
-      @data_import.table_name.should eq 'us_states'
+      # Verify the data import for the .carto file is not tied to a table
+      expect(@data_import.table_name).to be_nil
```
Comment (tylerparsons) [inline spec/connectors/importer_spec.rb:717]: Check `table_id` here as well.

File: spec/connectors/importer_spec.rb [hunk 7]
```diff
@@ -777,8 +799,8 @@
...
```
Comment (tylerparsons) [inline spec/connectors/importer_spec.rb:806]: Check `table_id` here as well.

File: spec/connectors/importer_spec.rb [hunk 8]
```diff
@@ -890,8 +927,8 @@
       expect(table_vis.export_geom).to be_true
       expect(table_vis.description).to eq('My dataset description')
 
-      # Check the user table was created - This always returns the first table
-      @data_import.table_name.should eq 'major_airports'
+      # Verify the data import for the .carto file is not tied to a table
+      expect(@data_import.table_name).to be_nil
```
Comment (tylerparsons) [inline spec/connectors/importer_spec.rb:935]: Check `table_id` here as well.

File: spec/connectors/importer_spec.rb [hunk 9]
```diff
@@ -912,5 +949,55 @@
...
+        append: false,
+        create_visualization: true,
+        synchronization_id: synchronization.id
+      )
+      @data_import.values[:data_source] = filepath
+
+      @data_import.run_import!
+      @data_import.success.should eq true
+      expect(synchronization.state).to eq 'success'
+      expect(synchronization.name).to be_nil
```
Comment (tylerparsons) [inline spec/connectors/importer_spec.rb:979]: Same checks for ids in this spec

File: spec/connectors/importer_spec.rb [hunk 10]
```diff
@@ -912,5 +949,55 @@
...
+
+      # Verify a synchronization record does NOT exists per dataset
+      sync = Carto::Synchronization.where(user_id: @user.id, name: 'local_data')
+      expect(sync.any?).to be_false
```
Comment (tylerparsons) [inline spec/connectors/importer_spec.rb]: Nice!

**Review by tylerparsons — CHANGES_REQUESTED** [review]
To clarify, the changes in `app/*` are all to restore original behavior for data imports, correct?  If so, the core logic looks good to me!

Just a few changes on test cases and comments.

**Comment by acarrera42** [comment]
Correct, the changes to table_registrar and user_table are to revert the previous changes.

File: app/models/data_import.rb (hunk 1 above)
Comment (acarrera42) [reply app/models/data_import.rb]: Correct, will update

File: services/importer/lib/importer/runner.rb (hunk 2 above)
Comment (acarrera42) [reply services/importer/lib/importer/runner.rb:281]: It is never added to the synchronizations in this change.  I noticed that if I manually added the dataset to a map (without previously connecting to the dataset) the synchronization record with the sql select does not appear either.

File: spec/connectors/importer_spec.rb (hunk 3 above)
Comment (acarrera42) [reply spec/connectors/importer_spec.rb:692]: the url is not set for the connector synchronization record.

**Comment by acarrera42** [comment]
Pushed changes requested.  Please re-review.  Thanks.

**Review by tylerparsons — APPROVED** [review]

**Pull request merged**